File: buildSrc/src/main/java/org/maplibre/nativeffi/gradle/MaplibreNativeCArtifact.java
package org.maplibre.nativeffi.gradle;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Arrays;
import java.util.List;
import java.util.Properties;

public final class MaplibreNativeCArtifact {
  private final File propertiesFile;
  private final Properties properties;

  public MaplibreNativeCArtifact(File propertiesFile) {
    this.propertiesFile = propertiesFile;
    this.properties = loadProperties(propertiesFile);
  }

  public File getPropertiesFile() {
    return propertiesFile;
  }

  public File getLibraryPath() {
    return new File(property("maplibreNativeC.libraryPath"));
  }

  public List<File> getIncludeDirs() {
    return pathList("maplibreNativeC.includeDirs").stream().map(File::new).toList();
  }

  public List<File> getLinkDirs() {
    return pathList("maplibreNativeC.linkDirs").stream().map(File::new).toList();
  }

  public List<File> getRuntimeLibraryDirs() {
    return pathList("maplibreNativeC.runtimeLibraryDirs").stream().map(File::new).toList();
  }

  public List<String> getLinkLibraries() {
    return pathList("maplibreNativeC.linkLibraries");
  }

  public List<String> getFrameworks() {
    return pathList("maplibreNativeC.frameworks");
  }

  private static Properties loadProperties(File propertiesFile) {
    var properties = new Properties();
    try (var input = new FileInputStream(propertiesFile)) {
      properties.load(input);
    } catch (IOException error) {
      throw new UncheckedIOException("Failed to read " + propertiesFile, error);
    }
    return properties;
  }

  private String property(String name) {
    return properties.getProperty(name, "");
  }

  private List<String> pathList(String name) {
    return Arrays.stream(property(name).split(File.pathSeparator, -1))
        .filter(value -> !value.isBlank())
        .toList();
  }
}

File: buildSrc/src/main/kotlin/org/maplibre/nativeffi/gradle/MaplibreNativeCArtifact.kt
package org.maplibre.nativeffi.gradle

import groovy.json.JsonSlurper
import java.io.File

class MaplibreNativeCArtifact(val installDir: File, val hostLibraryDirs: List<File>) {
  val libraryPath: File
    get() = runtimeLibraryDir.resolve(libraryFileName())

  val includeDirs: List<File>
    get() = listOf(installDir.resolve("include"))

  val linkDirs: List<File>
    get() = listOf(installDir.resolve("lib"))

  val runtimeLibraryDirs: List<File>
    get() = listOf(runtimeLibraryDir)

  val loaderLibraryDirs: List<File>
    get() = runtimeLibraryDirs + hostLibraryDirs

  val runtimeLinkLibraryDirs: List<File>
    get() = runtimeLibraryDirs + hostLibraryDirs

  val linkLibraries: List<String>
    get() = listOf("maplibre-native-c") + artifactStringList("staticLinkLibraries")

  val frameworks: List<String>
    get() = artifactStringList("staticLinkFrameworks")

  private val artifactMetadata: Map<*, *> by lazy {
    val descriptor = installDir.resolve("share/maplibre-native-c/artifact.json")
    if (descriptor.isFile) JsonSlurper().parse(descriptor) as Map<*, *> else emptyMap<Any, Any>()
  }

  private fun artifactStringList(name: String): List<String> =
    (artifactMetadata[name] as? List<*>)?.map { value ->
      require(value is String) { "$name in the native artifact descriptor must contain strings" }
      value
    } ?: emptyList()

  private val runtimeLibraryDir: File
    get() = installDir.resolve(if (targetIsWindows()) "bin" else "lib")

  private fun libraryFileName(): String =
    when {
      targetIsWindows() -> "maplibre-native-c.dll"
      targetIsMac() -> "libmaplibre-native-c.dylib"
      else -> "libmaplibre-native-c.so"
    }

  private fun targetIsMac(): Boolean = cargoBuildTarget()?.contains("apple-darwin") ?: isHostMac()

  private fun targetIsWindows(): Boolean =
    cargoBuildTarget()?.contains("windows-msvc") ?: isHostWindows()

  private fun cargoBuildTarget(): String? = System.getenv("CARGO_BUILD_TARGET")

  private fun isHostMac(): Boolean = System.getProperty("os.name").lowercase().contains("mac")

  private fun isHostWindows(): Boolean =
    System.getProperty("os.name").lowercase().contains("windows")
}
